### DISCUSSION TITLE
[Report] QIDI 6.08 Mono is 50% compatible with per layer settings

### DISCUSSION THREAD

**Discussion by human890209** [sn4k3/UVtools#260] [General]
Hi,
My printer is QIDI 6.08 Mono with original firmware V4.3.11_LCDKM / 1620 x 2560 / F2.21.

I tested the per-layer settings with the script example and some manual changes.
The ExposureTime and LightOffDelay work.

I found some differences:
1. When the layer's LiftHeight is set to 0mm, the firmware uses the General Properties' LiftHeight value. If General Properties's Lift Height value is set to 0, the firmware uses a built-in height value of 6~10mm
Maybe the walkaround is setting the layer's LiftHeight to 0.001mm manually.
2. Layer Index offset
Layer1 uses Layer2's LiftHeight value. Layer2 uses Layer3's. Layer3 uses Layer4's...
I don't know the offset number is constant 1 or the bottom layer count. Maybe I could modify the per-layer script sample to make a simple shift script.

Is there a way to install a generic chitu firmware(>= version 4.3.9) to my QIDI 6.08 Mono to get it 100% compatible?

**Comment by sn4k3** [comment]
It seens like QIDI have ctb v3 firmware compatible.
1. CTB and others can never use 0mm lift height or it will grab defaults, that's  why UVtools set a lift height of 0.1mm by **default** 
![UVtools_2021-08-08_14-20-17](https://user-images.githubusercontent.com/113281/128633375-477436a5-6d83-4a98-8187-0e9c715ee7c2.png)

2. I'm not sure what you mean, it must respect it own layer value and dont look one up...
3. Not a clue if you can install generic firmware, it requires someone to try to know if possible, but i will not risk it. Your firmware is already ctbv3 compatible

I will add that printer to the list 
Thanks

**Reply by human890209** [reply]
Update my research

0. The ExposureTime and LightOffDelay work for normal layers, bottom layers use BottomExposureTime and BottomLightOffDelay.
Set BottomLayerCount to 1 can maximize the layer settings control.
1. I tested and confirmed the MIN available LiftHeight for my printer is 0.001mm, and 0.0009mm failed.
2. I tested and confirmed the Lift&Retract mistake. It indeed uses the next layer's value, which could be an index mistake in firmware.
3.  I write a simple script to fix those.
[ScriptQIDI608MonoFix.zip](https://github.com/sn4k3/UVtools/files/6950917/ScriptQIDI608MonoFix.zip)

**Reply by sn4k3** [reply]
0. That is simillar to Photon Mono firmwares (crappy ones)
1. You should never use 0.001mm, printer accuracy are only up to 0.01mm and still risky as that is the limit. I use 0.1 to be safe anywhere
2. Crap firmwares.... What will happen on last layer? xD

**Reply by human890209** [reply]
1. Thanks for your 0.1mm tip. I found my printer Z accuracy is 0.00125mm. 0.1mm = 0.00125 * 80. Is 0.00125 a safe value?
2. The Last Layer seems to use itself, the Right one. xD
3. In my script:
```
// NoBottomLayer
if (NoBottomLayer.Value)
{
SlicerFile.BottomLayerCount = 0;
}
```
I set `SlicerFile.BottomLayerCount` to **0** and all the layer settings will be refreshed with General Properties.
Is there a way to disable [Propagate modifications to layers] in the script?
I tried
SlicerFile.SuppressRebuildProperties = true;
but it didn't work.

**Reply by sn4k3** [reply]
1. Do not use anything lower than 0.01mm as layer height or lifts, even for prints you should never go below 0.03mm as problems starts below that height are exponential.
3. Sure, use like this:

```
SlicerFile.SuppressRebuildPropertiesWork(() =>
{               
   SlicerFile.BottomLayerCount = 0;
});
```
That's the same as:
```
SlicerFile.SuppressRebuildProperties = true;
SlicerFile.BottomLayerCount = 0;
SlicerFile.SuppressRebuildProperties = false;
```

If that doesnt work is because you are updating other property that will rebuild latter

**Reply by human890209** [reply]
1. Did you mean 0.03mm? 0.3mm is 6 layers of 0.05mm which is a normal layer height.
2. It works! Thanks.

Update the fix script here:
[ScriptQIDI608MonoFix.zip](https://github.com/sn4k3/UVtools/files/6951006/ScriptQIDI608MonoFix.zip)

**Reply by sn4k3** [reply]
> Did you mean 0.03mm? 0.3mm is 6 layers of 0.05mm which is a normal layer height.

Yes (0.03mm), i have fixed.

> It works! Thanks.

😀

**Reply by human890209** [reply]
I record per layer settings test videos with my phone and watch it frame by frame with Resolve.
I found more about QIDI 6.08 Mono:
I tested with a crazy 10mm layer height and a normal 0.05 layer height.
The Lift and Retract split into 4 phases: Lift_1 > Lift_2 > Retract_1 > Retract_2
The distance of Lift_1 = 2mm(firmware const) + LayerHeight
The speed of Lift_1 = LiftSpeed value [~2mm, won't take much time with a slow LiftSpeed]
The distance of Lift_2 = LiftHeight - 2mm(firmware const)
The speed of Lift_2 = 120mm/min(firmware const)
The distance of Retract_1 = LiftHeight - 2mm(firmware const)
The speed of Retract_1 = RetractSpeed value
The distance of Retract_2 = 2mm(firmware const)
The speed of Retract_2 = 60mm/min(firmware const)

UVTools and my version of Chitubox can't edit LiftHeight2, LiftSpeed2...
I don't know if those Firmware Consts can be controlled by those X2 properties.

And the printing time calculation of Chitubox and UVTools LightOffDelay is not accurate.
But I guess a slow LiftSpeed and a super-fast RetractSpeed may get a good print in a short time.

**Reply by sn4k3** [reply]
New chitubox and ctbv4 can control that, but that seens a special firmware...
How do  you meassure the speed (firmware const?) Also do not confuse a second sequence with accelaration. near top or end the speed will decrease.

3rd plot:
![image](https://user-images.githubusercontent.com/113281/128638184-8abc0def-1bde-4276-be2c-431b306ec2d6.png)

**Reply by human890209** [reply]
I paste a ruler on my printer, measure the time by Video Editor's timestamp and frame. 30fps. frame by frame.
There indeed exists tween motion. But only 2-3 frames (0.1s). And I simply add up all the pause time around 0.5-1s.

**Reply by human890209** [reply]
Research Update：
The ExposureTime and LightOffDelay for Bottom Layers and Normal Layers are controlled by Global Properties if No Layer is modified. A modified layer sets RequireFullEncode to true and enables per-layer settings for the ctb file.

Updated Fix Script:
[ScriptQIDI608MonoFix.zip](https://github.com/sn4k3/UVtools/files/6951949/ScriptQIDI608MonoFix.zip)

**Reply by sn4k3** [reply]
`RequireFullEncode` is only to tell whatever a full save or partial save is needed. It will not set any special atribute to modify layers

**Reply by human890209** [reply]
I found if I opened an original A.ctb and modified some per-layer settings then saved it as B.ctb, the printer didn't use per-layer ExposureTime and LightOffDelay but the Global ones of B.ctb. I set RequireFullEncode to true with the script and saved it as C.ctb, the per-layer settings were used by the printer. Then I modified some per-layer settings saved it as D.ctb without RequireFullEncode enabled again. D.ctb's per-layer settings still worked.
I think a full save maybe override some data in the ctb file which enabled the per-layer settings for the printer.

**Reply by human890209** [reply]
The LiftSpeed and RetractSpeed data's index are also shifted. Update my fix script.
[ScriptQIDI608MonoFix.zip](https://github.com/sn4k3/UVtools/files/6953035/ScriptQIDI608MonoFix.zip)

**Reply by sn4k3** [reply]
Ah yes, when reencoding a special flag is set to the correct one to allow that.
I will put on normal save too

**Reply by human890209** [reply]
Now I know Lift & Retract is done before the Curing. So LightOffDelay is the one that should be shifted to the previous layer for my printer.
Update the script:
[V3Script_QIDI608MonoFix.zip](https://github.com/sn4k3/UVtools/files/6965117/V3Script_QIDI608MonoFix.zip)

**Reply by human890209** [reply]
I notice UVTools 2.18 support the TSMC. But the TSMC settings may be a bit different between V4 firmware and V3 firmware for my printer.

Here's my V4Script:
[ScriptQIDI608MonoFixV4.zip](https://github.com/sn4k3/UVtools/files/6985775/ScriptQIDI608MonoFixV4.zip)
I don't know other ctb V4 firmwares' behaviors.

The Per-Layer Settings of TSMC with UVTools 2.18 are welcomed, I need to update my script to override those data.

**Reply by sn4k3** [reply]
I see you having fun with scripting...
Note that "// Enable Per Layer Settings" fix is no longer need as UVtools will do that now

Also that if:

```
if (SlicerFile[layerIndex].IsBottomLayer)
ON: if (PropagateTSMC.Value)
```

Is not the best way...
Look here: `SlicerFile.GetInitialLayerValueOrNormal()`
eg: https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Layer/LayerManager.cs#L502

PS: Don't use 0.001f values UVtools round most to 2 decimals so use only down to 0.01f

**Reply by human890209** [reply]
Cool. Thanks for your tips.

**Reply by human890209** [reply]
I'm confused about the v3 ctb and v4 ctb file format.
My Chitubox slicer 1.9.0 basic outputs x.ctb files not x.v4.ctb files. It's ctbSlicerFile.HeaderSettings.Version is 4 and contains ctbSlicerFile.PrintParametersV4Settings.
What is the v4.ctb file format?

**Reply by sn4k3** [reply]
.v4.ctb is just a virtual extension i put on UVtools to force convert any file to V4. By default UVtools uses v3 to convert. But when loading it will just load whatever version the file contains.
Please read changelog to know what is happening

**Reply by human890209** [reply]
1. If I use SaveAs to create a new ctb file, the ctb version is the same as the original file I opened, all data is saved, right?
2. If I open a V2/V3 ctb file and use ConvertTo Chitubox CTBV4, it will override the Version Property to 4, override the EncryptMode to the right value, and add PrintParametersV4Settings data to let the V4 firmware printer print as a version converter, right?

Is changelog the one on the release page? I read that and get confused, and worried I don't save the ctb files in the right version.

**Reply by sn4k3** [reply]
1. Yes, **SaveAs** will keep original thing but with other name, but note that if a file need a full encode, if you name it xxx.v4.ctb it will force Version 4 because it will go with **Encode** method which does that check. If is a soft-save nothing is changed.
2. Correct. And from that on you can open it as normal .ctb on UVtools and it will be handled as v4. When loading files we ALWAYS have the version information so we don't need the virtual extension trick. But when creating files we have no way to user inform which version he want, so that virtual extension does the trick.

Explore code is the best way to understand what is happening: https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/FileFormats/ChituboxFile.cs#L1710

**Reply by human890209** [reply]
Thanks!
